Add "Cabal store maintenance" section to the manual (#1918)
well-typed/hs-bindgen: manual/Installation.md

  This is only needed when you want to use a version of LLVM/Clang that is
  not in your current `PATH`.

### Cabal store maintenance

The `libclang` bindings used by `hs-bindgen` are implemented in package
`libclang-bindings`.  When built, it is linked to the `libclang` and `libLLVM`
shared objects for the version of LLVM/Clang that you are using, and the built
`libclang-bindings` library is cached in the Cabal store.  Cabal does not
consider such system dependencies to determine when a package needs to be

> rebuilt, however, so you may continue to use a library built with one version

Reviewer: ```suggestion
rebuilt, however, so you might continue to use a library built an older version
```

the same minor version (and a newer patch version), then the link in the
cached `libclang-bindings` build is not broken.  The newer version *should*
still work, though we recommend rebuilding `hs-bindgen`/`libclang-bindings`
using the newer version.  In this case, `hs-bindgen` displays a warning like
the following, where the "compile time version" is the version of LLVM/Clang
used to to build the cached `libclang-bindings` library, and the "runtime
version" is the version for the shared object that is dynamically linked at
runtime.

```
[Warning] [HsBindgen] [boot] clang version mismatch:
  clang compile time version: 22.1.2
  clang runtime version:      22.1.3
```

If an upgrade installs a new major version and removes the older version, then
the link in the cached `libclang-bindings` build is broken.  In this case, the
dynamic library loader fails with an error like the following, preventing
`hs-bindgen` from running at all.

```
libclang.so.20.1: cannot open shared object file: No such file or directory
```

Cabal does not provide an easy way to resolve this issue.  Note that this is an
issue for any Haskell package that links to system libraries (without using

> `pkg-config`), but it is particularly problematic with LLVM/Clang because it

Reviewer: "without using `pkg-config`" -> "unless they provide `pkg-config` support, in which case cabal can track their dependencies just as it does for Haskell packages"

has frequent releases.  Currently, we suggest two ways to work around the issue.

The easiest way to force Cabal to rebuild out-of-date/broken packages is to
clear the cache.  You can determine your Cabal store directory by running
`cabal path --store-dir`.  The Cabal store has separate caches for each version
of GHC.  Recursively remove a GHC directory to clear the cache for that version.
Note that deleting just a package directory is not advised because it can break
the package database.  The downside to clearing your cache is that doing so may
result in time-consuming recompilation of many other packages.

> An alternative is to specify the LLVM/Clang version in `ghc-options`.  This

Reviewer: ```suggestion
Since we understand that clearing the cache might not be desirable, `hs-bindgen` offers a bespoke workaround, in the form of a compile-time setting specifying the LLVM/Clang version.  This
```

forces Cabal to distinguish `libclang-bindings` builds that link to different
versions, forcing a new build if one for the specified version is not already

> in the Cabal store.  Do this by setting the following in a `cabal.project.local`

Reviewer: ```suggestion
in the Cabal store.  Do this by adding the following to a `cabal.project.local`
```

  This is only needed when you want to use a version of LLVM/Clang that is
  not in your current `PATH`.

> ### Cabal store maintenance

Reviewer 2: Is this text also relevant for MacOS and Windows (I think so, right)? Because it is only a sub-section of the Linux section now. Maybe it should be moved out into its own top-level section

Author: Yes, it is indeed relevant for macOS and Windows.  With the recent change to only check the major/minor version for compatibility, ignoring the patch version, the issue is on Windows differs from Linux and macOS.  I have moved the documentation to the (new) `libclang-bindings` manual, referenced from the `hs-bindgen` manual.  The documentation is rewritten to make it clear how the issue manifests on the different OSes.
